Use SPRT as stopping condition

ThomasKolarik opened the pull request:
Fixes a TODO left by Gary.

The code is a modified version of what is found inside of CuteChess.

zz4032 commented:
Are the bounds elo0=-10 and elo1=10 used for stopping conditions? Are you sure you don't want elo0=**0** and elo1=10?

man4 commented:
Alpha and beta should be 0.05 (corresponding to 5% chance of Type I/II errors), not 0.5.

zz4032 commented:
Actually you can define alpha and beta as you wish depending on confidence that you want to have. 95% confidence (resulting in many games) may be draining to many resources for current fast improving rate of LCZero.
I would to set elo0=0, we don't want the new network to be worse, elo1 to a positive number and then test around with x=elo1 and y=alpha=beta.

zz4032 commented:
Alpha and beta define the confidence level and if you are fine with less games in a match and a lower confidence, you can increase alpha and beta.
We want to stop the match, if new network is better, so elo0=0.
We want the new network to pass, if it has 55% after 400 games (Elo +35) e.g. with typical draw percentage the results can be assumed W/L/D = 200/160/40.
Correct me if I'm wrong but with sprt calculator I get a pass result on conditions above with:
1) alpha=beta=0.2, Elo0=0, Elo1=14
2) alpha=beta=0.15, Elo0=0, Elo1=19
3) alpha=beta=0.1, Elo0=0, Elo1=32
The difference between the 3 examples is that with with option 1 generally less games will be necessary.
I have no idea how it was done for LeelaZero.
Additionally since the score after an sprt-test stoppage is biased, in case of a passed network the match should be replayed with a given number of games.

lucasart commented:
alpha and beta should be 0.05. Most importantly alpha! I don't know where you got these default values from…

lucasart commented:
@zz4032: what you propose gives 20% chance of accepting a 5 elo regression, and 50% chance of accepting a zero elo change. I don't think that's what we want…

zz4032 commented:
@ lucasart Which parameters would you suggest? We want to have a decision after about 400 games in a match on average.

jhellis3 commented:
Why that few of games? It is not as if they take a long time to complete at current settings.

zz4032 commented:
@jhellis3 My idea was only to replace the current match conditions based on fixed 400 games. Having longer matches is a valid point.

@lucasart Ok, if I shift the bounds to [0,10] with alpha=beta=0.2, I get 0,3% for accepting 5 Elo regression and 4,9% of accepting 0 Elo change.
Match results with alpha=beta=0.2 and bounds [0,10]:
```
Match ID  W      L       D      Elo   Games   Games (SPRT) 
27      220     198     65      +16     483   930 -> pass   1342
26      214     210     63      +3      487   5200 -> fail  7521
25      1       475     3       -910    479   48 -> fail    48
23      237     172     70      +47     479   311 -> pass   346
19      302     132     40      +130    474   124 -> pass   129
```
alpha=beta=0.05 will result in much more games, do we need that high level of confidence? For NN engine development it might be preferable to risk promoting a worse net than to spend endless games on matches.
Match results with alpha=beta=0.05 and bounds [0,10]:
```
Match ID  W      L       D      Elo   Games   Games (SPRT) 
27      220     198     65      +16     483   2840 -> pass
26      214     210     63      +3      487   15908 -> fail
25      1       475     3       -910    479   101 -> fail
23      237     172     70      +47     479   733 -> pass
19      302     132     40      +130    474   273 -> pass
```

lucasart commented:
@zz4032: if we're concerned that running matches takes too much resources, why do we run matches so often ? less frequent matches, using more rigorous match conditions (to guarantee that we only promote elo gains and don't risk too much missing small elo gains either) seem logical. I would suggest alpha=beta=5%, elo0=elo1=10.

zz4032 commented:
A new network is released every 4 hours currently. The main purpose of the tests currently is, as I understand, to catch networks that are much worse. Promoting a slightly worse network is not a disaster and adds diversity in games. Deepmind promoted every new network regardless of strength and claimed it was successful in the end.
The point of this pull request is to replace the current match conditions, which are not based on sprt and are definitely not optimal.
There are a lot of parameters and open questions to think about which are all connected between each other.
1) automatically promote new networks or do rigorous testing on each one of them
1) frequency of network releases
2) sprt bounds + confidence level in sprt tests
3) training window
4) when to increase layers
5) when/if increase/decrease number of playouts

vdbergh commented:
Since during testing the elo difference are typically quite large here (which is different from fishtest) it might be worthwhile to use the exact LLR instead of the approximation used in cutechess.  See here for 
an implementation:

http://hardy.uhasselt.be/Toga/SimpleSPRT.py

(this script also does discrete time correction and moreover it contains simulation code to validate the results).

zamar commented:
How about this as a first step:

The current passing condition is: after 400 games, determine if wins > losses.

Now there exists bounds alpha = 0.05, beta = 0.05, elo0 = -x, elo1 =+x which exhibits the exactly same behavior, but with only a fraction of games required (just determine the right x).

After that we can start fighting about better bounds, which is a difficult and non-scientific problem. Because it might actually be beneficial to accept slightly worse patches every now and then to allow NN weights to move around more (compare with simulated annealing).

zz4032 commented:
So for W/L/D 180/179/41 the sprt test shold result in a pass? I can't find any fitting parameters with my sprt script to make it pass with less than 1866 games. Am I doing something wrong?

alpha=beta=0.05, elo0=-100, elo1=100
```
          W      L       D      Elo    Score   Games   Games (SPRT) 
        180     179     41     +1     0.501   400       1866 -> pass
```

For the AlphaZero approach (score 55% after 400 games to pass) that would be basically similar to what I proposed above with alpha=beta=0.2 and x=5.5 (or alpha=beta=0.05 and x=11.6, which seems to be the same in that case).

alpha=beta=0.05, elo0=-11.6, elo1=11.6
```
          W      L       D      Elo    Score   Games   Games (SPRT) 
        200     160     40      +35     0.55   400       400 -> pass
```

Some previous matches:
```
Match ID  W      L       D      Elo   Games   Games (SPRT) 
27	220	198	65	+16	483   850 -> pass
26	214	210	63	+3	487   4773 -> pass
25	1	475	3	-910	479   44 -> fail
23	237	172	70	+47	479   283 -> pass
19	302	132	40	+130    474   113 -> pass
```

zamar commented:
And about the maximum game limit: It's a bad idea (unless set so high that we don't ever reach it in practice). Just makes result worse. If matches are too expensive on average, it's better to user looser SPRT bounds.

vdbergh commented:
Here is quick script that calculates the pass probabilities and expected running time of a SPRT test

http://hardy.uhasselt.be/Toga/sprta_logistic.py

It is a modification of a script I once made for fishtest. I did not validate the modified one by simulation but  the results seem sane.

vdbergh commented:
@zamar 

I propose to get rid of BayesElo. It causes endless confusion among users. Most of the developers at fishtest are still unaware of the difference between ordinary elo and BayesElo.

The use is BayesElo is a historical artifact as the standard SPRT allows only testing distributions that depend on a single parameter and in this case there are two parameters (elo and draw ratio). So in the beginning it seemed we needed a draw model to predict the draw ratio from the elo difference. 

But this was found to be inconvenient and currently both fishtest and cutechess simply measure the draw ration dynamically. So there is no point anymore of using a draw model.

While the implementation in fishtest and cutechess of the SPRT is a kludge and not entirely theoretically correct, it works very well for small elo differences.

However the correct theoretical approach to handle distributions that depend on more than one parameter is the GSPRT

http://stat.columbia.edu/~jcliu/paper/GSPRT_SQA3.pdf

This is what I have implemented in the script I posted above.

vdbergh commented:
@zamar 

Ah now I see that your question comes from the erroneous comment that the script uses BayesElo. This is not true. It uses standard elo. I have changed the comment.

zamar commented:
@vdbergh: Understood. Porting over the code you gave would be great: 
http://hardy.uhasselt.be/Toga/SimpleSPRT.py

It's just that someone needs to do it...

@zz4032: Playing around with @vdbergh tool bounds like [-20, 20] give an average run time of 457 games for 0 ELO patch (the most difficult case). 

joona@myfish:~$ python sprta_logistic.py -20 20 0.3 0

So it's fine. However just need to be prepared that some rare tests can take >1000 games (e.g. hard limit of 2000 games would be fine). Otherwise we should go with more relaxed bounds like [-25, 25]. Average run time of 292 games for 0 ELO patch (the most difficult case).

zz4032 commented:
I was about to ask one more question but now you already answered.

Just to make sure I understood. So if I want the most precise values and take any bound values (e.g. -20/20) and I go through all previous match results, run the script (adjusting draw ratio for each match) and the average running time of all of them is 100 games, then the bound were chosen correctly? One could do that iteratively.

zamar commented:
@zz4032: No, the bounds shouldn't be chosen iteratively... 1/4th was only a wild guess, it depends much on the data. The bounds should chosen based on the script @vdbergh provided:
http://hardy.uhasselt.be/Toga/sprta_logistic.py

As I said either [-25, 25] or [-20, 20] should work fine. Your results of 200 games and 250 games sound pretty realistic. You may want to play with @vdbergh 's script to draw your own conclusions.

vdbergh commented:
I cannot make a pull request since I have no time to set up the server locally but I translated the python script for executing the GSPRT into Go

http://hardy.uhasselt.be/Toga/SimpleSPRT.go

If there is interest in this then it should be easy to integrate.

Note this is my first foray into Go so the code is probably not very idiomatic.

zamar commented:
@ThomasKolarik: Would be interested in integrating @vdbergh's code? Instead of the current code borrowed from cutechess?

tokolarik commented:
Sure I will do this tomorrow evening.

ThomasKolarik commented:
Sure I will do this tomorrow evening.

glinscott commented:
Thanks for putting this together!!  So, after discussion with @Error323, what we'd like to go for is avoiding serious regressions, but allowing networks through, even if they are slight "regressions" in self-play.  Slight needs to be defined, but let's say if we keep our current ~500 match games, what sort of resolution could we hit?

zz4032 commented:
Have you thought about an additional fixed game match in case a network has succeeded in SPRT test? SPRT for the decision if to promote or not + fixed game test for Elo measurement afterwards. Because you can't use SPRT test results for serious elo estimations.
Btw in case you accept slight regressions, one could also drop the SPRT completely... The only point of SPRT was to save resources.

zamar commented:
@glinscott: One possibility that has been discussed here is [-20, 20] (Real ELO):
* Worst case average is 450 games. 
* -20 ELO patch has 5% chance of passing.
* 0 ELO patch has 50% chance of passing.
* +20 ELO patch has 95% chance of passing.

It's all guesswork... One could argue for [-30, +10] or [-10, +30] equally well... On one hand you want gating, on the other hand you want the weights to explore.

@zz4032: After you have set any kind of gating condition (like the current 400 games, wins > losses), your ELO measurements are completely skewed. SPRT is optimal for making binary decisions.

vdbergh commented:
I posted a new version of SimpleSPRT.go at the same link.

http://hardy.uhasselt.be/Toga/SimpleSPRT.go

The previous version could get into an infinite loop in case of extremely high elo differences.

I have now regularized te computation with a weak prior. The code is now more elegant as well.

ThomasKolarik commented:
Alright. I ported over the code. I hope I did the DB manipulation correctly. This was also my first delve into Golang so hopefully I didn't mess things up too badly.

zamar commented:
@ThomasKolarik: Thx.

BTW, GameCap also needs to be updated. It should be set so high that we won't practically ever reach it, but it's an extra catch if we end up in infinite loop due to a bug. Fox example you could set it to 5000.

lp-- commented:
@zamar 5000 limit  is clearly  an overkill, which negates all savings in games  from sprt.    Value of gating  is arguable (possibly negative), as  DeepMind went from gating  for AGZ  to no gating in AZ paper.   So there is no need to go crazy about it for the sake of mathematical accuracy. As @glinscott and @Error323 want to have 500 games  it is enough to use truncated sprt with 500 games set some elo0 and elo1 (-20, 20 is ok).  Calculations of accuracy and savings in the number of games may be done, and this is  it. 
For comparison Leela Zero uses sprt 0,35 soft limit around 400 games (or whatever played due to distribution of jobs), and if reached limit without passing (accept if more than 55%)

zamar commented:
- If the problem is too strict gating, then modify the bounds to be for example [-40, 0]. Almost everything will pass.
- If the problem is that the tests on average take too long, then use looser bounds e.g. [-25, 25].

Anything else (like extra truncation of the test) is just a bad statistical trade.

vdbergh commented:
@ThomasKolarik  Thanks a lot! At first sight it looks ok. I will be busy the whole day but will look at it tonight.

lp-- commented:
@zamar Seems to me you are refusing to accept real problem and try to fit everything in assumptions behind sprt model for which you know mathemtical solutions, and asking questions in terms of that model. Real world does not fit in there.

Here are  real problem:

-  For **whatever test** you are running after some number of games you always have trade-off between use of resources for test or for further training.  You can always say I don't care about this endless particular test (just make some decision) and spend resources for something else.  Such trade-off is not in sprt model premises.  If you want optimal solution put such premise in your model and solve it.

Here are other real problems not in the mathematical formulation of sprt model.

-  Games are sent to many workers at the same time.  Short games finish earlier and sprt can pass only based on improvements in the beginning of the game. Endgame may suffer much more and total elo result may be much worse than derived from games which caused sprt to pass.  This  happens with Leela Zero. How it influences learning I think is not known at this point, maybe someone can tell 
-  You need to have some tracking of performance, for people to see and have interest in project. You can have such estimate from sprt, but with small number of games it will be very inaccurate and biased. So this means more games for nets which are accepted than only from sprt test.

lp-- commented:
@zamar: Just for your info on statistical soundness

DeepMind went in AGZ paper with gating by fixed number of games test, and accepting those with >50%. Also they measured elo gain based on the result of the same test. This is statistically unsound. 

Leela Zero  after test passed sprt  is playing additional games up to 400  to measure elo gain. Sprt games are also included in this 400 games. This is likely give even more biased result than in AGZ.

Both had questionable solution from statistical point of view, both  seems to work though. There is trade-off between resources and mathematically sound results in models which you are able to solve.

zz4032 commented:
SPRT test conditions don't take into account that resources are not endless so a hard limit is reasonable. I would go for 2000 games limit. In Stockfish testing [-3,1] tests sometimes reach over 256000 games and are stopped for that same reason.

zamar commented:
@lp--:

1) It doesn't matter how long an individual test takes. What matters is how long one test takes *on average* and how accurate result we get. SPRT optimizes that. This is the perspective people don't seem to be able to understand. Why is it worse to run "19x100 game tests and 1x1000 game tests" than "20 x 300 games" test?

2) There will always be variables beyond our control, e.g. hardware between testing machines are not identical. The fact that there is some noise by testing set up, is not a good excuse to introduce even more noise on purpose.

3) About statistical soundness: What AGZ did is statistically unsound. What LZ does is even more unsound. They make progress, because NN backpropagation works. Even if the gating condition was "discard 50% of the new weights randomly", they would still make great progress. Even if you put wrong kind of oil in your car, you can probably drive it just fine. But why would you do that?

@zz4032: As long as the stop limit is high enough that it's reached very infrequently, it's fine. 2000 might just be high enough for [-20, 20]. (BTW, In fishtest when 256000 games is reached, operators have the power to bump up the limit to continue the test).

lp-- commented:
@zamar 

> It doesn't matter how long an individual test takes. What matters is how long one test takes on average and how accurate result we get. SPRT optimizes that. This is the perspective people don't seem to be able to understand. Why is it worse to run "19x100 game tests and 1x1000 game tests" than "20 x 300 games" test?

Take case when sprt test  'as is' is perfectly applicable. You have conveyor line and quality control. You **really want** to know answer with set confidence level.  Here you care only about minimizing **average** length of trials  of  individual tests, and don't care about length of every individual test.

In our case when test survives some sufficiently large number of games **we don't really care** that we know with set confidence level whether to accept null hypothesis or reject. We can just make decision, it will be  good enough. And we can save resources for more important tasks. This is the difference.  

I did not do calculations, but I think those long tests contribute  large chunk to increase average number of games, so it will be noticeable.

.

zamar commented:
@lp--: Yes, conveyor line and quality control is an interesting example. So far I've assumed that we only want to optimize the average length of trials and ignore the standard deviation.

Now there is plenty of math for truncated SPRT tests where you bring the standard deviation down, but the cost to pay is the longer length of trials on average. Note that "truncated SPRT test" in the literature doesn't just mean "use normal SPRT test + stop it after X games", but there are different, more conservative formulas for bounds.

If we go back to conveyor line where you do the test under limited time constraints, "truncated SPRT test" makes perfect sense. The reason that we didn't introduce it for Fishtest was that we saw no point in making the tests to run longer on average. Rather if one test runs longer than the others we just tolerate it...

And to be honest, I see no point in introducing truncated SPRT tests for LCZ either. You decide the filtering condition, and want to get the result as quickly as possible (on average). If one test runs longer than others you just tolerate it (rather than speeding up this one specific run at the cost of making all the other tests to run longer).

lp-- commented:
> About statistical soundness: What AGZ did is statistically unsound. What LZ does is even more unsound. They make progress, because NN backpropagation works. Even if the gating condition was "discard 50% of the new weights randomly", they would still make great progress. Even if you put wrong kind of oil in your car, you can probably drive it just fine. But why would you do that?

My guess that DeepMind just did not realize that. They likely had enough resources to do everything right.  But LZ don't have enough resources,  so 'good enough' is perfectly fine decision here.

For lczero this 5000 games or whatever number may look like not a big deal now, but when net will get bigger and number of games played  decrease by a lot, they will  be important

zamar commented:
@lp--: If we hit the 5000 games test once per year, it doesn't matter in the big picture. It's just all about the average case...

vdbergh commented:
> I did not do calculations, but I think those long tests contribute large chunk to increase average number of games, so it will be noticeable.

You do not seem to understand the concept of an expectation value. 

Mathematics tells you what the expected number of games of an SPRT is. Some tests will take longer and some tests will be less long. But the average length will be equal to the computed expected length. The SPRT  even has the property that the expected length is optimal under certain conditions.

You cannot cheat statistics by truncating long tests. Yes this will reduce the expected number of games but it will also degrade the characteristics (probabilities of false positives and negatives) of the test.

If you want a faster SPRT then just do one with less stringent characteristics. This script does the calculations.  

http://hardy.uhasselt.be/Toga/sprta_logistic.py

lp-- commented:
> You do not seem to understand the concept of an expectation value.
> 
> Mathematics tells you what the expected number of games of an SPRT is. Some tests will take longer and some tests will be less long. But the average length will be equal to the computed expected length. The SPRT even has the property that the expected length is optimal under certain conditions.
> 
> You cannot cheat statistics by truncating long tests. Yes this will reduce the expected number of games but it will also degrade the characteristics (probabilities of false positives and negatives) of the test.
> 
> If you want a faster SPRT then just do one with less stringent characteristics. This script does the calculations.
> 
> http://hardy.uhasselt.be/Toga/sprta_logistic.py

Michel,  i can assure you that not only mathematicians understand what expected value is :)
What I was arguing here is that sprt model premises does not exactly fit the problem we are trying to solve.  Think about it, maybe there is better model.

vdbergh commented:
@ThomasKolarik

As far as I can tell the port seems ok! Thanks again.

One comment: it seems to me that in sprt.go you should delete "main()" and the simulation code. It is not useful in a package.

zamar commented:
@glinscott: Do we want to keep the -50 ELO limit (basically always pass except serious regressions)?

Basically these are the steps we want to take:
1) Define the pass limit, e.g. -50 ELO. Effectively this means that -50 ELO patch has 50% chance of passing.
2) Decide how much resources we want to spend in the testing (average: games/test). If we want to spend on average something like 300 games per test, then probably something like [-70, -30] for SPRT boundaries should work just fine.

ThomasKolarik commented:
@vdbergh Good catch. 

I changed the GetStatus method to pass by reference as well but after doing a bit of googling I am not so sure that is better.

http://goinbigdata.com/golang-pass-by-pointer-vs-pass-by-value/

I also updated the max number of games to 1000 as well. With -20, 20 bounds we should really not ever hit this.

Let me know if there need to be any modifications or if anybody else finds some bugs. I'll be happy to fix them!

zamar commented:
@ThomasKolarik: 2000 is a reasonable minimum for max number of games. Personally I would go with 5000, but other people have issues with that.

vdbergh commented:
@ThomasKolarik  I agree with @Zamar that an upper bound which is almost never hit (like 5000 for a test with a resolution of 40) is best. The people that want a lower upper bound are incorrectly implying that the resource consumption of the SPRT is determined by this upper bound. In reality it is determined by the expected number of games (which will be arranged to be around 400).

An upper bound which is hit frequently drastically alters the characteristics of the SPRT and they become substantially more difficult to compute.

At fishtest they went to an upper bound of 1 million.

vdbergh commented:
About elo estimates: You can get statistically correct elo estimates from SPRT tests. In fact this script does precisely that  http://hardy.uhasselt.be/Toga/sprt_elo.py (this is for the cutechess/fishtest SPRT, a slight modification is necessary for the GSPRT as in the current PR).

However as pointed out by @zamar (and others) the gating process itself creates bias, no matter what type of test is used (you only measure winners). This bias appears unavoidable without investing more resources (that is: running an independent test to measure progress).

With the current plan to essentially abolish gating except for large regressions, this issue would be mitigated however.

vdbergh commented:
I created once again a new version which does overshoot correction in a more theoretically correct way (and the code is again more elegant as well).  

http://hardy.uhasselt.be/Toga/sprt.go

Note: overshoot correction compensates for the fact that when the SPRT stops, the LLR will have slightly overshot  the boundaries of the continuation region, due to the fact that time is discrete. For short test (where the LLR makes large jumps) this can substantially influence the error probabilities.

Simulation shows that the corrected version is accurate for very large elo differences.  

When (and if...) the current PR is committed I will issue a new one with the update sprt package.

jkiliani commented:
If you are dealing with overshoot, are you also taking the queue buffer into account? Currently we produce 700-800 match games instead of 400 all the time, for the simple reason that the server keeps queuing them until 400 are reached...

vdbergh commented:
> If you are dealing with overshoot, are you also taking the queue buffer into account? Currently we produce 700-800 match games instead of 400 all the time, for the simple reason that the server keeps queuing them until 400 are reached...

If you have more games coming in after the SPRT has reached a decision you can use them to obtain a more unbiased elo measurement... Statistically the acceptance/rejection decision should be uncoupled from the elo measurement.

PS. The overshoot I was talking about is different. The SPRT as defined by Wald assumes that the tests stops when the LLR is exactly on one of the boundaries of the continuation region but this cannot happen since the LLR changes in discrete steps. This effect is only measurable for short tests (like the matches).

zz4032 commented:
@zamar 
> After you have set any kind of gating condition (like the current 400 games, wins > losses), your ELO measurements are completely skewed. SPRT is optimal for making binary decisions.
If one runs a second match (with fixed number of games) as I proposed after the promoting decision has been made (based on sprt match), there would be no skew in the progress chart.
